# Laboratorio 2 - Reglas de Asociación y Patrones Secuenciales

## Grupo 1 - Paula Hípola Gómez, José Ignacio Navas Sanz y Belén Ortega Pérez

## Entregable 1 - Reglas de Asociación

### 1.1 (1 punto) Estudia brevemente el dataset (número de usuarios, número de atributos, categorías más frecuentes para cada atributo/columna, etc.) y realiza el preprocesamiento para generar el dataset necesario para aplicar el algoritmo Apriori a este conjunto de datos.

In [24]:
%config IPCompleter.greedy=True
%reset
#Importamos las librerías necesarias
#Numpy
import numpy as np 
#Pandas
import pandas as pd
#Scikit-Learn
import sklearn as sk
#Matplotlib
import matplotlib.pyplot as plt

Once deleted, variables cannot be recovered. Proceed (y/[n])? y


### Cargamos los datos y los visualizamos

In [25]:
#Importamos el CSV
dataset = pd.read_csv('radio.csv')

#Mostramos los datos
dataset.head()

user                   artist sex  country
0     1    red hot chili peppers   f  Germany
1     1  the black dahlia murder   f  Germany
2     1                goldfrapp   f  Germany
3     1         dropkick murphys   f  Germany
4     1                 le tigre   f  Germany

In [26]:
print('La dimensión del dataset es:',dataset.shape)

La dimensión del dataset es: (289955, 4)


Tenemos 289955 instancias y 4 atributos, para cada una de nuestras instancias. Analizando más profundamente el dataset, está compuesto con las interacciones de cada usuario que ha escuchado al menos a un artista. Vamos a ver cuantos valores tenemos actualmente en cuanto a usuarios, artistas y demás atributos y luego procederemos a identificar en el dataset valores duplicados y nulos que dificultan la comprensión de los datos y procederemos a su eliminación.

In [27]:
numUsuarios = dataset.groupby('user').count().shape[0]
print('Hay',numUsuarios,'usuarios.')
numArtistas = dataset.groupby('artist').count().shape[0]
print('Hay',numArtistas,'artistas.')
numSexos = dataset.groupby('sex').count().shape[0]
print('Hay',numSexos,'géneros.')
numPaises = dataset.groupby('country').count().shape[0]
print('Hay',numPaises,'países.')

Hay 15000 usuarios.
Hay 1004 artistas.
Hay 2 géneros.
Hay 159 países.


Como podemos ver, hemos agrupado todos los usuarios, artistas, géneros y países en sus respectivas categorías para comprender mejor el dataset planteado, ahora procederemos a preprocesar y limpiar los datos. Si analizamos este dataset, vemos que tenemos 15000 usuarios si bien este no es un dato real puesto que hay saltos en el ID de usuario. Esto se debe a que tenemos los datos de los usuarios que han escuchado al menos 1 artista por lo que no podemos sacar una conclusión del número exacto de usuarios, si no de que hay 15000 usuarios que han escuchado al menos a un artista.

### Limpieza de datos

In [28]:
#Eliminación de nulos
dataset = dataset.dropna()
print('Tamaño sin nulos',dataset.shape)

#Eliminación de duplicados
dataset = dataset.drop_duplicates()
print('Tamaño sin duplicados',dataset.shape)
dataset

Tamaño sin nulos (289955, 4)
Tamaño sin duplicados (289953, 4)


user                   artist sex  country
0           1    red hot chili peppers   f  Germany
1           1  the black dahlia murder   f  Germany
2           1                goldfrapp   f  Germany
3           1         dropkick murphys   f  Germany
4           1                 le tigre   f  Germany
...       ...                      ...  ..      ...
289950  19718                bob dylan   f   Canada
289951  19718                   pixies   f   Canada
289952  19718                the clash   f   Canada
289953  19718     a tribe called quest   f   Canada
289954  19718                radiohead   f   Canada

[289953 rows x 4 columns]

In [29]:
dataset['country'].describe()

count            289953
unique              159
top       United States
freq              59558
Name: country, dtype: object

In [30]:
dataset['artist'].describe()

count        289953
unique         1004
top       radiohead
freq           2704
Name: artist, dtype: object

Tenemos que hacer un nuevo dataset para analizar estos datos con el algoritmo Apriori, para ello vamos a construir un dataset con un listado de ususarios con todos los artistas que ha escuchado...

In [31]:
dagrupado = dataset.groupby('user')['artist'].apply(list)
dfAgrupado = pd.DataFrame(dagrupado)
dfAgrupado

artist
user                                                    
1      [red hot chili peppers, the black dahlia murde...
3      [devendra banhart, boards of canada, cocorosie...
4      [tv on the radio, tool, kyuss, dj shadow, air,...
5      [dream theater, ac/dc, metallica, iron maiden,...
6      [lily allen, kanye west, sigur rós, pink floyd...
...                                                  ...
19713  [armin van buuren, above & beyond, atb, ferry ...
19714  [misfits, type o negative, arch enemy, red hot...
19715  [abba, james blunt, jason mraz, amy winehouse,...
19717  [marilyn manson, beyoncé, madonna, t.a.t.u., k...
19718  [beirut, of montreal, black flag, the new porn...

[15000 rows x 1 columns]

Pese a tener los datos del país del usuario además de su género, nos damos cuenta de que estos datos carecen de valor y no aportan nada al objetivo que perseguimos. Que no es más que la adquisición de los conjuntos de grupos de música más frecuentes para la obtención de reglas de asociación que nos permita el estudio de estos grupos y sus similitudes por el cual se produzcan estos agrupamientos.

### 1.2 (4 puntos) Aprende a manejar la librería mlextend y su implementación del algoritmo Apriori.
#### 1.2.1 Obtén los itemsets frecuentes para k=1. Para esto es necesario obtener el soporte de los itemset. Crea una función que utilice métodos de la librería mlextend de manera que dado un itemset devuelva su soporte.



In [32]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(dagrupado).transform(dagrupado)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

...and you will know us by the trail of dead   2pac  3 doors down  \
0                                             False  False         False   
1                                             False  False         False   
2                                             False  False         False   
3                                             False  False         False   
4                                             False  False         False   
...                                             ...    ...           ...   
14995                                         False  False         False   
14996                                         False  False         False   
14997                                         False  False         False   
14998                                         False  False         False   
14999                                         False  False         False   

       30 seconds to mars    311  36 crazyfists     44  50 cent  \
0                   False  False          False  False    False   
1                   False  False          False  False    False   
2                   False  False          False  False    False   
3                   False  False          False  False    False   
4                   False  False          False  False    False   
...                   ...    ...            ...    ...      ...   
14995               False  False          False  False    False   
14996               False  False          False  False    False   
14997               False  False          False  False    False   
14998               False  False          False  False    False   
14999               False  False          False  False    False   

       65daysofstatic  [unknown]  ...  wilco  within temptation  \
0               False      False  ...  False              False   
1               False      False  ...  False              False   
2               False      False  ...  False              False   
3               False      False  ...  False              False   
4               False      False  ...  False              False   
...               ...        ...  ...    ...                ...   
14995           False      False  ...  False              False   
14996           False      False  ...  False              False   
14997           False      False  ...  False              False   
14998           False      False  ...  False              False   
14999           False      False  ...  False              False   

       wolfgang amadeus mozart  wu-tang clan  yann tiersen  yeah yeah yeahs  \
0                        False         False         False            False   
1                        False         False         False            False   
2                        False         False         False            False   
3                        False         False         False            False   
4                        False         False         False            False   
...                        ...           ...           ...              ...   
14995                    False         False         False            False   
14996                    False         False         False            False   
14997                     True         False         False            False   
14998                    False         False         False            False   
14999                    False         False         False             True   

       yellowcard  yo la tengo  zero 7  Édith piaf  
0           False        False   False       False  
1           False        False   False       False  
2           False        False   False       False  
3           False        False   False       False  
4           False        False   False       False  
...           ...          ...     ...         ...  
14995       False        False   False       False  
14996       False        False   False       False  
14997       False        False   False       False  
14998

In [33]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(df, min_support=0.015, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets1 = frequent_itemsets[frequent_itemsets['length'] == 1]

In [34]:
print('El valor medio de soporte para itemset 1 es:',frequent_itemsets1["support"].mean())

El valor medio de soporte para itemset 1 es: 0.032591526778577164


In [35]:
print(frequent_itemsets1.to_string())

      support                        itemsets  length
0    0.022733                          (2pac)       1
1    0.030933                  (3 doors down)       1
2    0.032800            (30 seconds to mars)       1
3    0.021800                       (50 cent)       1
4    0.036867                     ([unknown])       1
5    0.036733              (a perfect circle)       1
6    0.018533          (a tribe called quest)       1
7    0.028133                          (abba)       1
8    0.061400                         (ac/dc)       1
9    0.033800                     (aerosmith)       1
10   0.028533                           (afi)       1
11   0.069600                           (air)       1
12   0.019267                          (akon)       1
13   0.027733             (alanis morissette)       1
14   0.025733               (alice in chains)       1
15   0.024133                   (alicia keys)       1
16   0.019600                 (alkaline trio)       1
17   0.022533               

#### 1.2.2. Crea una función que devuelva los itemsets frecuentes candidatos y su soporte para k>=2



In [36]:
frequent_itemsets2 = frequent_itemsets[frequent_itemsets['length'] == 2]
frequent_itemsets2

support                          itemsets  length
417  0.016267          (a perfect circle, tool)       2
418  0.015133              (iron maiden, ac/dc)       2
419  0.017467             (ac/dc, led zeppelin)       2
420  0.022067                (ac/dc, metallica)       2
421  0.015267               (ac/dc, pink floyd)       2
..        ...                               ...     ...
727  0.021733                 (the beatles, u2)       2
728  0.016333            (the smiths, the cure)       2
729  0.018267          (the kooks, the killers)       2
730  0.018600        (the killers, the strokes)       2
731  0.015067  (the killers, the white stripes)       2

[315 rows x 3 columns]

#### 1.2.3. Crea una función que repita el paso 1.2.2. hasta que no se generen nuevos itemsets frecuentes



In [15]:
frequent_itemsets3 = frequent_itemsets[frequent_itemsets['length'] == 3]
print(frequent_itemsets3.to_string())

       support                                              itemsets  length
1959  0.009733                            (coldplay, air, radiohead)       3
1960  0.009133                      (air, massive attack, radiohead)       3
1961  0.009333                         (air, radiohead, the beatles)       3
1962  0.009467                 (arcade fire, radiohead, the beatles)       3
1963  0.009133                (coldplay, bloc party, arctic monkeys)       3
1964  0.009467             (the killers, bloc party, arctic monkeys)       3
1965  0.011600                      (coldplay, arctic monkeys, muse)       3
1966  0.013067                 (coldplay, arctic monkeys, radiohead)       3
1967  0.009533               (coldplay, arctic monkeys, the beatles)       3
1968  0.012333               (coldplay, the killers, arctic monkeys)       3
1969  0.009133        (franz ferdinand, the killers, arctic monkeys)       3
1970  0.011000                     (arctic monkeys, radiohead, muse)       3

In [13]:
#print(df.to_string())

#### 1.2.4. Crea una función que muestre todas las reglas posibles con su confianza



#### 1.2.5. Lista todas las reglas que sean de alta confianza (p.ej. con valor >=0.5)



#### 1.2.6. Crea funciones que permitan lo siguiente:
<ul>
    <li><b>* devolver todas las reglas que contengan un antecesor dado</b></li>
    <li><b>* devolver todas las reglas que tengan una confianza mayor (o igual) que un umbral mínimo dado</b></li>
    <li><b>* devolver todas las reglas que tengan un lift en una de las tres franjas de estudio (under 1, over 1, =1)</b></li>
</ul>



#### 1.2.7. Utiliza al menos dos representaciones gráficas para mostrar las reglas obtenidas e interpretar los datos